### PR TITLE
Use IsIncluded only on folder entries

### PR DESCRIPTION
**Pull request by kewillford**
@wilbaker these are the changes to use only the `IsIncluded` on a folder.

### CODE REVIEW

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs [hunk 1]
```diff
@@ -691,17 +675,18 @@ private static List<ProjectedFileInfo> ConvertToProjectedFileInfos(SortedFolderE
             {
                 childEntry = sortedFolderEntries[i];
 
-                if (childEntry.IsIncluded)
+                if (childEntry.IsFolder)
```
Comment (derrickstolee) [inline GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Perhaps `if ((childEntry as FolderData)?.IsIncluded == true)`? Maybe that's too terse and the nested `if` is better.

File: GVFS/GVFS.UnitTests/Virtualization/Projection/SortedFolderEntriesTests.cs [hunk 2]
```diff
@@ -98,7 +98,7 @@ public void AddItemAtTheBeginning()
         {
             SortedFolderEntries sfe = SetupDefaultEntries();
             LazyUTF8String name = ConstructLazyUTF8String("((shouldbeatthestart");
-            sfe.AddFolder(name);
+            sfe.AddFile(name, new byte[20]);
```
Comment (wilbaker) [inline GVFS/GVFS.UnitTests/Virtualization/Projection/SortedFolderEntriesTests.cs]: Why was this test (and `AddItemAtEnd`) changed to insert a file rather than a folder?

File: GVFS/GVFS.UnitTests/Virtualization/Projection/SortedFolderEntriesTests.cs [hunk 3]
```diff
@@ -117,15 +117,45 @@ public void ValidateOrderOfDefaultEntries()
...
         {
             SortedFolderEntries sfe = SetupDefaultEntries();
             LazyUTF8String name = ConstructLazyUTF8String("IsIncludedFalse");
-            sfe.AddFile(name, new byte[20], isIncluded: false);
-            sfe.TryGetValue(name, out FolderEntryData folderEntryData).ShouldBeTrue();
-            folderEntryData.ShouldNotBeNull();
-            folderEntryData.IsFolder.ShouldBeFalse();
-            folderEntryData.IsIncluded.ShouldBeFalse();
+            sfe.GetOrAddFolder(new[] { name }, partIndex: 0, parentIsIncluded: false, rootSparseFolderData: new SparseFolderData());
+            ValidateFolder(sfe, name, isIncludedValue: true);
```
Comment (wilbaker) [inline GVFS/GVFS.UnitTests/Virtualization/Projection/SortedFolderEntriesTests.cs]: Is `isIncludedValue` true because `rootSparseFolderData` is empty?

I had some trouble parsing the name of this test, perhaps something like `AddFolderWhenSparseFolderDataIsEmtpy` or `FoldersShouldBeVisibleWhenSparseFolderDataIsEmpty`?

File: GVFS/GVFS.UnitTests/Virtualization/Projection/SortedFolderEntriesTests.cs [hunk 4]
```diff
@@ -117,15 +117,45 @@ public void ValidateOrderOfDefaultEntries()
...
+
+        [TestCase]
+        public void AddFolderWhereParentIncludedIsTrueAndChildIsNotIncluded()
+        {
+            SortedFolderEntries sfe = SetupDefaultEntries();
+            LazyUTF8String name = ConstructLazyUTF8String("ChildNotIncluded");
+            SparseFolderData sparseFolderData = new SparseFolderData();
+            sparseFolderData.Children.Add("Child", new SparseFolderData());
+            sfe.GetOrAddFolder(new[] { name }, partIndex: 0, parentIsIncluded: true, rootSparseFolderData: sparseFolderData);
+            ValidateFolder(sfe, name, isIncludedValue: false);
```
Comment (wilbaker) [inline GVFS/GVFS.UnitTests/Virtualization/Projection/SortedFolderEntriesTests.cs]: Is there a test to handle the case where a parent folder is recursively included?

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs [hunk 5]
```diff
@@ -1008,51 +965,50 @@ private FolderData AddFileToTree(GitIndexEntry indexEntry)
...
 
-        /// <summary>
-        /// Finds the FolderEntryData for the path provided
-        /// </summary>
-        /// <param name="pathParts">Path to desired entry</param>
-        /// <param name="folderEntryData">Out: FolderEntryData for pathParts</param>
-        /// <returns>True if the path could be found in the tree, and false otherwise</returns>
-        /// <remarks>If the root folder is desired, the pathParts should be an empty array</remarks>
-        private bool TryGetFolderEntryDataFromTree(LazyUTF8String[] pathParts, out FolderEntryData folderEntryData)
+        private bool TryGetFolderDataFromTreeUsingPath(string folderPath, out FolderData folderData)
```
Comment (wilbaker) [inline GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Could you add a comment for this method explaining what it does?

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs [hunk 6]
```diff
@@ -1062,14 +1018,13 @@ private bool TryGetFolderEntryDataFromTree(LazyUTF8String[] pathParts, out Folde
         /// childPathParts.
```
Comment (wilbaker) [inline GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Minor:  Above comment is no longer accurate:

```
        /// Finds the FolderEntryData at the specified depth of the path provided.  If depth == pathParts.Length
        /// TryGetFolderEntryDataFromTree will find the FolderEntryData specified in pathParts.  If
        /// depth is less than pathParts.Length, then TryGetFolderEntryDataFromTree will return an ancestor folder of
        /// childPathParts.
```

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs [hunk 7]
```diff
@@ -1501,7 +1456,7 @@ private string GetNewProjectedShaForPlaceholder(string path)
             HashSet<string> existingPlaceholders)
         {
             bool foundFolder = this.TryGetOrAddFolderDataFromCache(relativeFolderPath, out FolderData folderData);
-            if (!foundFolder || !folderData.IsIncluded)
+            if (!foundFolder)
```
Comment (wilbaker) [inline GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Why was the `!folderData.IsIncluded` check removed here?

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs [hunk 8]
```diff
@@ -485,7 +469,7 @@ public virtual bool IsPathProjected(string virtualPath, out string fileName, out
                 childName: fileName,
```
Comment (wilbaker) [inline GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Maybe add a comment above the call to `GetProjectedFolderEntryData` that mentions `null` will be returned if a path's parent folder is not included.

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs (hunk 1 above)
Comment (wilbaker) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Out of curiosity, how would you expect the performance of `as` to compare to the virtual `IsFolder` call?

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs [hunk 9]
```diff
@@ -428,45 +428,29 @@ public virtual void GetFileTypeAndMode(string filePath, out FileType fileType, o
...
             try
             {
-                this.GetChildNameAndParentKey(virtualPath, out string fileName, out string parentKey);
-                FolderEntryData data = this.GetProjectedFolderEntryData(
-                    blobSizesConnection: null,
-                    childName: fileName,
-                    parentKey: parentKey);
-
-                if (data != null && !data.IsIncluded)
+                if (this.TryGetFolderDataFromTreeUsingPath(virtualPath, out FolderData folderData) &&
```
Comment (wilbaker) [inline GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: It looks like with this change we'll be skipping the lookup in the cache that would happen in `GetProjectedFolderEntryData`.  Do we know at this point that this folder won't be in the cache?

(I am guessing that because we're in `TryAddSparseFolder` we'd know the folder is not in the cache?)

Adding a comment about out expectations of the cache would be helpful.

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs [hunk 10]
```diff
@@ -428,45 +428,29 @@ public virtual void GetFileTypeAndMode(string filePath, out FileType fileType, o
...
-            if (data == null)
-            {
-                return PathSparseState.NotFound;
-            }
-            else if (data.IsIncluded)
-            {
-                return PathSparseState.Included;
-            }
-            else
+            if (this.TryGetFolderDataFromTreeUsingPath(virtualPath, out FolderData folderData))
```
Comment (wilbaker) [inline GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Similar question as in `TryAddSparseFolder`, why was this changed from `GetProjectedFolderEntryData` to `TryGetFolderDataFromTreeUsingPath`?

File: GVFS/GVFS.UnitTests/Virtualization/Projection/SortedFolderEntriesTests.cs (hunk 2 above)
Comment (kewillford) [reply GVFS/GVFS.UnitTests/Virtualization/Projection/SortedFolderEntriesTests.cs]: Simply because it is an easier call.  The `AddFolder` method was removed and now there is
```
public FolderData GetOrAddFolder(
    LazyUTF8String[] pathParts,
    int partIndex,
    bool parentIsIncluded,
    SparseFolderData rootSparseFolderData)
```

Which is more setup work to call and checking the order shouldn't matter whether it is a file or a folder

File: GVFS/GVFS.UnitTests/Virtualization/Projection/SortedFolderEntriesTests.cs (hunk 3 above)
Comment (kewillford) [reply GVFS/GVFS.UnitTests/Virtualization/Projection/SortedFolderEntriesTests.cs]: >Is isIncludedValue true because rootSparseFolderData is empty?

Yes because if the root sparse folder data doesn't have any children then everything should be included.

I will update the name of the test.

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.SortedFolderEntries.cs [hunk 11]
```diff
@@ -88,27 +88,57 @@ public void Clear()
...
+                            }
+                        }
+                    }
+                    else
+                    {
+                        isIncluded = false;
+                    }
+                }
+
+                return this.InsertFolder(pathParts[partIndex], ~index, isIncluded: isIncluded);
```
[the remark this answers is not in the document]
Comment (kewillford) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.SortedFolderEntries.cs]: If the entry is not found the index is set the in the complement of the value where it should be inserted.  This is so there only needs to be one call to look something up instead of looking it up to discover it isn't found and looking up the index where it should be inserted when it isn't found.

File: GVFS/GVFS.UnitTests/Virtualization/Projection/SortedFolderEntriesTests.cs (hunk 2 above)
Comment (wilbaker) [reply GVFS/GVFS.UnitTests/Virtualization/Projection/SortedFolderEntriesTests.cs]: Ok thanks!

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.SortedFolderEntries.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (wilbaker) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.SortedFolderEntries.cs]: Thanks that makes sense

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs (hunk 1 above)
Comment (kewillford) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Running a simple test and `as` was in most cases about 2x slower but I had to up the iterations in the test to at least a million before it was even taking 1 ms so unless we are using it in a tight loop I'm sure other code will be a bottleneck before this.

Comment (derrickstolee) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Well, thanks for checking the perf! That would have been a mistake to listen to me.

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs (hunk 10 above)
Comment (kewillford) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: `GetProjectedFolderEntryData` will not return entries that are not included and we we need to be able to return `Excluded` entries.

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs (hunk 9 above)
Comment (kewillford) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: This is for the create new folder scenario and we checked the sparse state and got back excluded in order to get to this code and excluded folders do not get added to the cache.  I will add some comments.

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs (hunk 7 above)
Comment (kewillford) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: `TryGetOrAddFolderDataFromCache` will return false when it is not included
```
if (!this.projectionFolderCache.TryGetValue(folderPath, out folderData))
{
	if (!this.TryGetFolderDataFromTreeUsingPath(folderPath, out folderData) ||
		!folderData.IsIncluded)
	{
		folderData = null;
		return false;
	}

	this.projectionFolderCache.TryAdd(folderPath, folderData);
}

return true;
```

File: GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs [hunk 12]
```diff
@@ -1612,7 +1582,7 @@ private FSResult RemoveFolderPlaceholderIfEmpty(IPlaceholderData placeholder)
             string childName;
             string parentKey;
             this.GetChildNameAndParentKey(placeholder.Path, out childName, out parentKey);
-            PathSparseState sparseState = this.GetPathSparseState(placeholder.Path);
+            PathSparseState sparseState = this.GetFolderPathSparseState(parentKey);
```
Comment (wilbaker) [inline GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Can this line be removed completely?  It looks like `sparseState` is only getting used for this check:

`sparseState != PathSparseState.Included`

And as best I can tell `this.TryGetSha` will return false for folders that are not included (it uses `GetProjectedFolderEntryData`).  If we can remove this call it will save us a lot of walking the tree in the more common scenario that placeholders on disk are still included.

Comment (kewillford) [reply GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs]: Looks like it.  I will give it a try and see.
